### PR TITLE
Initial commit of web encoders

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
This is the first commit in a series to make web encoders part of core libraries.
We have had some discussions with the ASP.NET team, and we concluded that the encoders
are quite general purpose, and should be made part of the core.

This commit includes a subset of Microsoft.Framework.WebEncoders.Core that we deemed general purpose.
It excludes some APIs related to dependency injection and some ASP.NET specific helpers.

The only changes in the moved code are:
a) new namespace and assembly name (System.Text.Encodings.Web)
b) changes to project files required by the CoreFx repo build system
c) changes to embedded resource paths (as the directory structure has changed)

One caveat, some tests in the test project are disabled. They do pass, but the test harness logger
fails as it cannot output some exotic characters that are passed to InlineData (and so printed in the log).
The root cause of this is most probably in XmlEscape routine in
https://github.com/xunit/xunit/blob/40e153fe983977be8fe009384fae5c377bc557ff/src/common/XmlTestExecutionVisitor.cs.
I will fix this problem in a subsequent commit.

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
Please only merge it after @Eilon signs off on this. Thanks!

**Comment by KrzysztofCwalina** [comment]
test it please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  <PropertyGroup>
+    <MinimumVisualStudioVersion>10.0</MinimumVisualStudioVersion>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{1DD0FF15-6234-4BD6-850A-317F05479554}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>System.Text.Encodings.Web</RootNamespace>
+    <AssemblyName>System.Text.Encodings.Web</AssemblyName>
+    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
```
Comment (stephentoub) [inline src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj]: There's a lot more stuff in here than in our other .csproj files; I'd suggest comparing and culling.  As part of that, should TargetFrameworkVersion and TargetFrameworkProfile be set like this?

Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj]: Thanks. I simplified it.

**Comment by akoeplinger** [comment]
Do all of the added classes really belong into System.Text.Encodings.**Web**? e.g. I'd imagine CodePointFilter is very useful outside of a web context as well.

**Comment by stephentoub** [comment]
I think something got messed up in the process of your addressing feedback: the latest commit says "Showing  2,113 changed files  with 25,053 additions and 463 deletions."

**Comment by KrzysztofCwalina** [comment]
Hmmm, I will retract the PR and resubmit after I am back from vacations. The codebase is using C# 6.0 and although it's compiling locally, it is massively failing the CI :-(

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UnicodeRanges.generated.cs [hunk 2]
```diff
@@ -0,0 +1,1406 @@
...
+{
+    public static partial class UnicodeRanges
+    {
+        /// <summary>
+        /// A <see cref="UnicodeRange"/> corresponding to the 'Basic Latin' Unicode block (U+0000..U+007F).
+        /// </summary>
+        /// <remarks>
+        /// See http://www.unicode.org/charts/PDF/U0000.pdf for the full set of characters in this block.
+        /// </remarks>
+        public static UnicodeRange BasicLatin => Volatile.Read(ref _basicLatin) ?? CreateRange(ref _basicLatin, first: '\u0000', last: '\u007F');
```
Comment (sharwell) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UnicodeRanges.generated.cs]: :bulb: There is no need to make this a volatile read. You can use just this:

```
... => _basicLatin ?? ...
```

**Comment by Eilon** [comment]
BTW the copyright+license headers of the files need to change.

Also, please keep in mind that there are some pending changes in the original repo, so we'll need to port those changes at some point.

**Comment by KrzysztofCwalina** [comment]
@akoeplinger, you are right. some of these don't belong here. I will be working with @ellismg on subsequent refactoring.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UnicodeRanges.generated.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/UnicodeRanges.generated.cs]: yes, I will remove these.

**Comment by KrzysztofCwalina** [comment]
@Eilon I changed the headers to the standard corefx headers. And I will merge the changes from original repo as they get committed.

**Pull request closed without merging**